### PR TITLE
[Workspace] Use NX as workspace manager

### PR DESCRIPTION
**Pull request by Hyperkid123**
### Changes
- remove lerna
- use NX as workspace manager
  - builds
  - linting
  - testing
  - playground app 

### TODO
- [x] cypress tests
- [x] verify output

### Discovered issues

Some packages had circular dependencies. This was hidden due to usage of `@ts-ignore` comments to prevent the TSC from failing the builds. These conflicts must be removed. The new build does not allow circular dependencies.

PR is against an nx branch, there is still the release process to be updated to GH actions and NX plugins and documentation on how to use the new workspaces manager and some workflow restrictions to streamline the development. These will be done separately.

This is still WIP.

### CODE REVIEW

**Comment by florkbr** [comment]
> Some packages had circular dependencies. This was hidden due to usage of @ts-ignore comments to prevent the TSC from failing the builds. These conflicts must be removed. The new build does not allow circular dependencies.

@Hyperkid123 I'd really like to incorporate [madge](https://www.npmjs.com/package/madge) into our build/development tooling - maybe as an optional step you can enable in your fec.config or package.json. I don't think this is the first nor the last repo that has circular dependencies with ts-ignores.

**Comment by florkbr** [comment]
@Hyperkid123 when I check this out locally I see git now thinks some JS/TS files are now untracked that are important:
```bash
➜  frontend-components git:(bc911de8) ✗ git status
HEAD detached at Hyperkid123/nx
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	packages/advisor-components/RebootRequired/
	packages/advisor-components/ReportDetails/
	packages/advisor-components/RuleDescription/
        ...
	packages/charts/Donut/
	packages/charts/Gauge/
	packages/charts/Matrix/
	packages/charts/Pie/
	...
	packages/chrome/ChromeContext/
	packages/chrome/ChromeProvider/
	...
	packages/components/Ansible/
	packages/components/AsyncComponent/
	packages/components/Battery/
	...
	(more below; truncated for readability)
```

Did those move somewhere else?

**Comment by florkbr** [comment]
@Hyperkid123 the build and lint steps seem fine:
### Build lerna (before)
```bash
lerna success run Ran npm script 'build' in 14 packages in 19.6s:
lerna success - @redhat-cloud-services/frontend-components-advisor-components
lerna success - @redhat-cloud-services/frontend-components-charts
lerna success - @redhat-cloud-services/chrome
lerna success - @redhat-cloud-services/frontend-components
lerna success - @redhat-cloud-services/frontend-components-config-utilities
lerna success - @redhat-cloud-services/frontend-components-config
lerna success - @redhat-cloud-services/frontend-components-notifications
lerna success - @redhat-cloud-services/frontend-components-pdf-generator
lerna success - @redhat-cloud-services/frontend-components-remediations
lerna success - @redhat-cloud-services/rule-components
lerna success - @redhat-cloud-services/frontend-components-testing
lerna success - @redhat-cloud-services/frontend-components-translations
lerna success - @redhat-cloud-services/tsc-transform-imports
lerna success - @redhat-cloud-services/frontend-components-utilities
```

### Build nx (after)
```bash
> @redhat-cloud-services/source@0.0.1 build
> nx run-many --target=build --exclude=demo,@redhat-cloud-services/frontend-components-pdf-generator,@redhat-cloud-services/frontend-components-charts


   ✔  nx run @redhat-cloud-services/types:build (954ms)
   ✔  nx run @redhat-cloud-services/frontend-components-testing:build:bundles (1s)
   ✔  nx run @redhat-cloud-services/tsc-transform-imports:build (959ms)
   ✔  nx run @redhat-cloud-services/frontend-components-testing:transform:scss (268ms)
   ✔  nx run @redhat-cloud-services/frontend-components-config-utilities:build (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-utilities:build:bundles (2s)
   ✔  nx run @redhat-cloud-services/frontend-components-testing:build:styles (360ms)
   ✔  nx run @redhat-cloud-services/frontend-components-testing:build:packages (297ms)
   ✔  nx run @redhat-cloud-services/frontend-components-utilities:build:packages (312ms)
   ✔  nx run @redhat-cloud-services/frontend-components-utilities:transform:scss (316ms)
   ✔  nx run @redhat-cloud-services/frontend-components-utilities:build:styles (424ms)
   ✔  nx run @redhat-cloud-services/frontend-components-utilities:build (2ms)
   ✔  nx run @redhat-cloud-services/frontend-components-translations:build:bundles (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-translations:transform:scss (304ms)
   ✔  nx run @redhat-cloud-services/chrome:build:bundles (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-translations:build:styles (353ms)
   ✔  nx run @redhat-cloud-services/chrome:build:packages (281ms)
   ✔  nx run @redhat-cloud-services/chrome:build (746μs)
   ✔  nx run @redhat-cloud-services/frontend-components-testing:build (809μs)
   ✔  nx run @redhat-cloud-services/frontend-components-translations:build:packages (274ms)
   ✔  nx run @redhat-cloud-services/frontend-components-translations:build (1ms)
   ✔  nx run @redhat-cloud-services/frontend-components-config:build (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-executors:build (1s)
   ✔  nx run @redhat-cloud-services/frontend-components:build:bundles (5s)
   ✔  nx run @redhat-cloud-services/frontend-components:transform:scss (464ms)
   ✔  nx run @redhat-cloud-services/frontend-components:build:packages (481ms)
   ✔  nx run @redhat-cloud-services/frontend-components:build:styles (956ms)
   ✔  nx run @redhat-cloud-services/frontend-components:build (1ms)
   ✔  nx run @redhat-cloud-services/frontend-components-remediations:build:bundles (2s)
   ✔  nx run @redhat-cloud-services/rule-components:build:bundles (3s)
   ✔  nx run @redhat-cloud-services/frontend-components-notifications:build:bundles (2s)
   ✔  nx run @redhat-cloud-services/rule-components:transform:scss (321ms)
   ✔  nx run @redhat-cloud-services/frontend-components-notifications:transform:scss (301ms)
   ✔  nx run @redhat-cloud-services/rule-components:build:packages (305ms)
   ✔  nx run @redhat-cloud-services/frontend-components-notifications:build:styles (421ms)
   ✔  nx run @redhat-cloud-services/frontend-components-remediations:transform:scss (298ms)
   ✔  nx run @redhat-cloud-services/frontend-components-remediations:build:styles (349ms)
   ✔  nx run @redhat-cloud-services/frontend-components-advisor-components:build:bundles (2s)
   ✔  nx run @redhat-cloud-services/frontend-components-remediations:build:packages (285ms)
   ✔  nx run @redhat-cloud-services/frontend-components-advisor-components:transform:scss (291ms)
   ✔  nx run @redhat-cloud-services/rule-components:build:styles (463ms)
   ✔  nx run @redhat-cloud-services/frontend-components-advisor-components:build:packages (296ms)
   ✔  nx run @redhat-cloud-services/rule-components:build (915μs)
   ✔  nx run @redhat-cloud-services/frontend-components-remediations:build (572μs)
   ✔  nx run @redhat-cloud-services/frontend-components-notifications:build:packages (278ms)
   ✔  nx run @redhat-cloud-services/frontend-components-notifications:build (857μs)
   ✔  nx run @redhat-cloud-services/frontend-components-advisor-components:build:styles (401ms)
   ✔  nx run @redhat-cloud-services/frontend-components-advisor-components:build (772μs)

——————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 NX   Successfully ran target build for 14 projects and 34 tasks they depend on (12s)
```

### Lint lerna (before)
```bash
✖ 236 problems (0 errors, 236 warnings)
  0 errors and 92 warnings potentially fixable with the `--fix` option
```

### Lint nx (after)
```bash
➜  frontend-components git:(bc911de8) npm run lint

> @redhat-cloud-services/source@0.0.1 lint
> nx run-many -t lint --exclude=@redhat-cloud-services/frontend-components-pdf-generator,@redhat-cloud-services/frontend-components-charts


   ✔  nx run @redhat-cloud-services/types:lint (2s)
   ✔  nx run @redhat-cloud-services/tsc-transform-imports:lint (2s)
   ✔  nx run @redhat-cloud-services/frontend-components-config-utilities:lint (2s)
   ✔  nx run demo:lint (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-config:lint (2s)
   ✔  nx run @redhat-cloud-services/frontend-components-utilities:lint (2s)
   ✔  nx run @redhat-cloud-services/chrome:lint (2s)
   ✔  nx run @redhat-cloud-services/frontend-components-executors:lint (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-testing:lint (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-translations:lint (1s)
   ✔  nx run @redhat-cloud-services/frontend-components:lint (4s)
   ✔  nx run @redhat-cloud-services/frontend-components-remediations:lint (1s)
   ✔  nx run @redhat-cloud-services/frontend-components-advisor-components:lint (2s)
   ✔  nx run @redhat-cloud-services/rule-components:lint (2s)
   ✔  nx run @redhat-cloud-services/frontend-components-notifications:lint (2s)

——————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 NX   Successfully ran target lint for 15 projects (10s)
```

However I think `test:unit` and `test:component` still need work. `test:unit` doesn't run any tests for me locally and `test:component` throws errors

**Comment by Hyperkid123** [comment]
@florkbr can you share the error for the cypress tests? We will probably have to sit down to see what is going on with your unit tests. It is running for me.

File: package.json [hunk 1]
```diff
@@ -1,152 +1,163 @@
...
+    "vitest": "^1.3.1",
+    "webpack-cli": "^5.1.4",
+    "whatwg-fetch": "^3.6.2"
+  },
+  "scripts": {
+    "nx": "nx",
+    "build": "nx run-many --target=build --exclude=demo,@redhat-cloud-services/frontend-components-pdf-generator,@redhat-cloud-services/frontend-components-charts",
+    "lint": "nx run-many -t lint --exclude=@redhat-cloud-services/frontend-components-pdf-generator,@redhat-cloud-services/frontend-components-charts",
+    "serve:demo": "nx run demo:serve",
+    "test:unit": "nx run-many -t test --exclude=demo",
```
[the remark this answers is not in the document]
Comment (Hyperkid123) [reply package.json:146]: Oh yeah, I fully intent on doing that. I'd like to do that in a follow where we would be setting up the CI environment.

**Comment by Hyperkid123** [comment]
> However I still see the issue switching to this branch from the upstream nx branch. If we removed those packages intentionally while we migrate to NX, should we do a git rm so they don't have to be manually removed in our local envs?

We will have to do some "sync" of packages before we merge the PR there are changes already and I am also thinking about discontinuing the pdf and charts packages.

**Review by florkbr — APPROVED** [review]

**Pull request merged**